cfe_internal/update/update_policy.cf

```diff
@@ -299,14 +297,112 @@ bundle agent cfe_internal_update_policy_cpv
+      # there will be no cf-serverd listening to serve files yet.
+    enterprise_edition.!(am_policy_hub.bootstrap_mode)::  # ENT-6840
+      "cfe_internal_update_cmdb_data_consumption";
+}
+
+bundle agent cfe_internal_update_cmdb_data_distribution
+# @brief Ensure data is ready for agents to download
+{
+  vars:
+    policy_server.enterprise_edition::
+
+      # The API response for host specific data from cmdb tells us the timestamp of the last data change
+      # We store this timestamp and use it for the next request.
+      "_cmdb_next_request_state_file"
+        string => "$(sys.statedir)/cmdb_next_request_from.dat";
+
+      # If we have the timestamp from a previous response we use it, else we start from 0
+      "_cmdb_previous_next_request_from"
+        string => readfile( $(_cmdb_next_request_state_file), inf ),
+        if => regline( "^\d+$", $(_cmdb_next_request_state_file) );
+
+      "_cmdb_previous_next_request_from"
+        string => "0",
+        unless => regline( "^\d+$", $(_cmdb_next_request_state_file) );
+
+      # We need a script to call that should return the API response
+      "_get_cmdb_data_bin" string => "$(sys.bindir)/get_cmdb.php";
+      "_get_cmdb_data_cmd" string => "/var/cfengine/httpd/php/bin/php $(_get_cmdb_data_bin) $(_cmdb_previous_next_request_from)";
+
+      # We call the script and we pass it the timestamp from the prior call
+      "_get_cmdb_data_response"
+        string => execresult( $(_get_cmdb_data_cmd), useshell ), # This would be execresult()
+        if => fileexists( $(_get_cmdb_data_bin) );
+
+      "_get_cmdb_data_response_d"
+        data => mergedata('$(_get_cmdb_data_response)');
+
+      # So that we can write a JSON file for each host we get the indicies of data in the response
+      "_i" slist => getindices( "_get_cmdb_data_response_d[data]");
+
+      # We need to store the timestamp from the most recent change so that we can use that as a starting point for future requests.
+      "_next_request_from"
+        string => "$(_get_cmdb_data_response_d[meta][cmdb_epoch])";
+
+  files:
+      # "$(_get_cmdb_data_cmd)" perms => m( 700 );
+
+@if minimum_version(3.18)
+      # This functionality is only present on 3.18.0+ Enterprise hubs, and this
+      # promise uses the /content/ attribute which was first introduced in
+      # 3.16.0.
+
+      # If the next request state file doesn't exist, we seed one with 0, the
+      # lowest epoch value possible. because we populate variables from this
+      # file content.
+      "$(_cmdb_next_request_state_file)"
+        content => "0$(const.n)",
+        handle => "cmdb_data_change_next_seed",
+        unless => fileexists("$(_cmdb_next_request_state_file)" );
+@endif
+
+      # Write out the data for each host that had a data change
+      "$(sys.workdir)/cmdb/$(_i)/host_specific.json"
+        create => "true", # CFE-2329, ENT-4792
+        template_data => mergedata("_get_cmdb_data_response_d[data][$(_i)]" ),
```

Comment:
`mergedata()` shouldn't be needed here either.

Reply:
Hum, I think it is.

Reply:
I think it is needed here. 

```
 #+begin_src cfengine3 :include-stdlib t :log-level info :exports both
   bundle agent __main__
   {
     vars:
         "_d" data => '{
         "1": { "color": "red" },
         "2": { "color": "blue" },
         "3": { "color": "green" }
         }';

         "_i" slist => getindices( _d );

     files:

         "/tmp/$(_i).json"
           create => "true",
           template_data => parsejson( "_d[$(_i)]"),
           template_method => "inline_mustache",
         edit_template_string => string_mustache( "{{$-top-}}", "_d[$(_i)]" );

   }
#+end_src 

#+RESULTS:
#+begin_example
   error: Error parsing JSON expression '_d[1]': Unable to parse json data as object, unrecognized token beginning entry
   error: Error parsing JSON expression '_d[2]': Unable to parse json data as object, unrecognized token beginning entry
   error: Error parsing JSON expression '_d[3]': Unable to parse json data as object, unrecognized token beginning entry
   error: Error parsing JSON expression '_d[1]': Unable to parse json data as object, unrecognized token beginning entry
   error: No template data for the promise '/tmp/1.json'
   error: Errors encountered when actuating files promise '/tmp/1.json'
   error: Error parsing JSON expression '_d[2]': Unable to parse json data as object, unrecognized token beginning entry
   error: No template data for the promise '/tmp/2.json'
   error: Errors encountered when actuating files promise '/tmp/2.json'
   error: Error parsing JSON expression '_d[3]': Unable to parse json data as object, unrecognized token beginning entry
   error: No template data for the promise '/tmp/3.json'
   error: Errors encountered when actuating files promise '/tmp/3.json'
   error: Error parsing JSON expression '_d[1]': Unable to parse json data as object, unrecognized token beginning entry
   error: Error parsing JSON expression '_d[2]': Unable to parse json data as object, unrecognized token beginning entry
   error: Error parsing JSON expression '_d[3]': Unable to parse json data as object, unrecognized token beginning entry
   error: Error parsing JSON expression '_d[1]': Unable to parse json data as object, unrecognized token beginning entry
   error: Error parsing JSON expression '_d[2]': Unable to parse json data as object, unrecognized token beginning entry
   error: Error parsing JSON expression '_d[3]': Unable to parse json data as object, unrecognized token beginning entry
#+end_example

```